From ae08f21db322ab0d08841bf6a95a30ec88afd9e4 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Fri, 22 May 2026 09:22:48 +0800
Subject: [PATCH] mtk: hostapd: add valid_link in sta_info's mld_info

Store the STA's valid links when handling the association request.
The valid_links will be used in the following commit (for TTLM).

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ieee802_11.c     | 17 +++++++++++++++++
 src/ap/ieee802_11_eht.c |  2 ++
 src/ap/sta_info.c       |  1 +
 src/ap/sta_info.h       |  1 +
 4 files changed, 21 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 5fa3733e3..86331a651 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -3732,6 +3732,7 @@ static void handle_auth(struct hostapd_data *hapd,
 			ap_sta_set_mld(sta, true);
 			sta->mld_assoc_link_id = link_id;
 			sta->mld_assoc_sta = sta;
+			sta->mld_info.valid_links = BIT(link_id);
 
 			/*
 			 * Set the MLD address as the station address and the
@@ -5489,10 +5490,12 @@ int hostapd_process_assoc_ml_info(struct hostapd_data *hapd,
 	int ret = 0;
 #ifdef CONFIG_IEEE80211BE
 	unsigned int i;
+	u16 valid_links;
 
 	if (!hostapd_is_multiple_link_mld(hapd))
 		return 0;
 
+	valid_links = BIT(sta->mld_assoc_link_id);
 	for (i = 0; i < MAX_NUM_MLD_LINKS; i++) {
 		struct hostapd_data *bss = NULL;
 		struct mld_link_info *link = &sta->mld_info.links[i];
@@ -5531,6 +5534,20 @@ int hostapd_process_assoc_ml_info(struct hostapd_data *hapd,
 				    LINK_PARSE_ASSOC, offload,
 				    set_beacon))
 				ret = -1;
+			else
+				valid_links |= BIT(bss->mld_link_id);
+		}
+	}
+
+	if (valid_links) {
+		struct hostapd_data *bss;
+
+		for_each_mld_link(bss, hapd) {
+			struct sta_info *lsta;
+
+			lsta = ap_get_sta(bss, sta->addr);
+			if (lsta)
+				lsta->mld_info.valid_links = valid_links;
 		}
 	}
 #endif /* CONFIG_IEEE80211BE */
diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index e8482b77f..e32b7df8f 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -1907,6 +1907,7 @@ void hostapd_link_reconf_resp_tx_status(struct hostapd_data *hapd,
 			link->resp_sta_profile = NULL;
 			link->resp_sta_profile_len = 0;
 			link->valid = false;
+			other_sta->mld_info.valid_links &= ~BIT(link_id);
 		}
 		wpa_auth_set_ml_info(assoc_sta->wpa_sm,
 				     assoc_sta->mld_assoc_link_id,
@@ -2040,6 +2041,7 @@ hostapd_ml_process_reconf_link(struct hostapd_data *hapd,
 		_link = &other_sta->mld_info.links[link_id];
 		_link->valid = true;
 		_link->status = WLAN_STATUS_SUCCESS;
+		other_sta->mld_info.valid_links |= BIT(link_id);
 		os_memcpy(_link->local_addr, other_hapd->own_addr, ETH_ALEN);
 		os_memcpy(_link->peer_addr, link_addr, ETH_ALEN);
 	}
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index c04c86c23..12e80eb9a 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -2293,6 +2293,7 @@ static void hostapd_reset_assoc_sta(struct hostapd_data *hapd, struct sta_info *
 			li->resp_sta_profile = NULL;
 			li->resp_sta_profile_len = 0;
 		}
+		link_sta->mld_info.valid_links &= ~BIT(hapd->mld_link_id);
 	}
 
 	for_each_mld_link(link_bss, hapd) {
diff --git a/src/ap/sta_info.h b/src/ap/sta_info.h
index cc165bb88..2bbb0b4ac 100644
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -73,6 +73,7 @@ struct pending_eapol_rx {
 
 struct mld_info {
 	bool mld_sta;
+	u16 valid_links;
 
 	struct ml_common_info {
 		u8 mld_addr[ETH_ALEN];
-- 
2.45.2

